### DISCUSSION TITLE
SQL Lab Row Limit feature request / Bug fix

### DISCUSSION THREAD

**Discussion by zuzana-vej** [apache/superset#19709] [Ideas]
Currently when result in SQL Lab has less than 10k records, it shows the total number of results. But if more than 10k results got returned, frontend only displays 10k results. It prompts user to download CSV to see all results. In some cases it is critical for users to see the number of returned rows (those that would be downloaded within the CSV) - to (1) see if it matches their expectation (sometimes they might realize sql has issue) and (2) to be able to verify if they download the CSV that it got downloaded correctly with all the results.
Currently only way to verify is to modify the SQL query and run select count(......). This is not a good workaround for less technical people who might just be reusing any saved query. 

Currently:
Less than 10k results:
<img width="421" alt="Screen Shot 2022-04-13 at 3 43 47 PM" src="https://user-images.githubusercontent.com/61221714/163295445-0b0a5a5e-8f25-4b25-a4a6-f24595e6c24a.png">

More than 10k results: Displays 10k, but I don't know what's the total:
<img width="1072" alt="Screen Shot 2022-04-13 at 6 15 20 PM" src="https://user-images.githubusercontent.com/61221714/163295390-512b8a56-a210-4258-88fe-8db88714145b.png">

Proposed solution:
Always displayed number of rows returned by backend, even when frontend limits display to 10k.

**Comment by zuzana-vej** [comment]
@yousoph @michael-s-molina thoughts on this feature? I can see how this is very frustrating experience for users for some reason I thought we addressed it a while back

**Reply by yousoph** [reply]
The suggestion makes sense to me! Maybe it's something we can display in the 10k result text  message somewhere

**Reply by zuzana-vej** [reply]
I think it could be something like this: 
10000 rows displayed, our of 459089 rows returned by the query. 

I tried to look briefly but didn't easily find if/where we return the total row count from backend (otherwise changing the message is simple). So I was wondering if someone with more familiarity with this section of code knows.

**Comment by rumbin** [comment]
On a complex query this may make a huge difference in terms of compute costs, though.

If no `order by` is specified, the database can stop processing after it computed 10 k records.
If we want to know the total number of records returned, the database needs to actually perform the computation on _all_ records, even if we then only output 10 k records.

Imagine that our query might easily return several tens or hundreds of billions of rows. Then our query costs (and duration!) can increase by a factor of, say, 1e5.

While I do understand that it would be very useful, at times, to get an information on the total number of records, we need to keep in mind that there are actual big data use cases where this approach is not feasible.

**Comment by kgeis** [comment]
I think of DBeaver, where if the row count hits the limit, it says for example "200+", and there is a button next to the row count to "Calculate total row count" if you need it.
